independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanStream.java

```diff
@@ -0,0 +1,156 @@
+package io.quarkus.arc.processor;
+
+import java.lang.annotation.Annotation;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.Type;
+
+/**
+ * Convenient {@link Stream} wrapper that can be used to filter a set of beans.
+ * <p>
+ * This object is stateful and cannot be reused. After a terminal opration is performed, the underlying stream is considered
+ * consumed, and can no longer be used.
+ * <p>
+ * This construct is not threadsafe.
+ */
+public final class BeanStream implements Iterable<BeanInfo> {
+
+    private Stream<BeanInfo> stream;
+
+    public BeanStream(Collection<BeanInfo> beans) {
+        this.stream = Objects.requireNonNull(beans, "Beans collection is null").stream();
+    }
+
+    public BeanStream withScope(Class<? extends Annotation> scope) {
+        return withScope(DotName.createSimple(scope.getName()));
+    }
+
+    public BeanStream withScope(DotName scopeName) {
+        stream = stream.filter(bean -> bean.getScope().getDotName().equals(scopeName));
+        return this;
+    }
+
+    public BeanStream withBeanType(Class<?> beanType) {
+        return withBeanType(DotName.createSimple(beanType.getName()));
+    }
+
+    public BeanStream withBeanType(DotName beanTypeName) {
+        stream = stream.filter(bean -> bean.getTypes().stream().anyMatch(t -> t.name().equals(beanTypeName)));
+        return this;
+    }
+
+    public BeanStream withBeanType(Type beanType) {
+        stream = stream.filter(bean -> bean.getTypes().stream().anyMatch(t -> t.equals(beanType)));
+        return this;
+    }
+
+    public BeanStream withBeanClass(Class<?> beanClass) {
+        return withBeanClass(DotName.createSimple(beanClass.getName()));
+    }
+
+    public BeanStream withBeanClass(DotName beanClass) {
+        stream = stream.filter(bean -> bean.getBeanClass().equals(beanClass));
+        return this;
+    }
+
+    public BeanStream withQualifier(Class<? extends Annotation> qualifier) {
+        return withQualifier(DotName.createSimple(qualifier.getName()));
+    }
+
+    public BeanStream withQualifier(DotName qualifierName) {
+        stream = stream.filter(bean -> bean.getQualifiers().stream().anyMatch(q -> q.name().equals(qualifierName)));
+        return this;
+    }
+
+    public BeanStream withName(String name) {
+        stream = stream.filter(bean -> name.equals(bean.getName()));
+        return this;
+    }
+
+    public BeanStream withIdentifier(String id) {
```

Comment:
Most of these methods could use at least a one-liner javadoc.
I know it's pretty obvious to us, but for example many users will have zero clue what identifier/name actually is for a bean.